### ISSUE TITLE
Removing static calls

### ISSUE THREAD

**Issue by mickadoo** [civicrm/org.civicrm.api4#63]
Static calls are fine for something that has no internal state, the CRM_Utils_Array class, for example, is very helpful. However by using a static based approach for the API we make it hard to test new classes, and harder to keep track of dependencies.

For example take a class that tries to find duplicate contacts. If I want to test the logic for finding duplicates I need to create them in the database if it fetches contacts via static calls. This adds a lot of setup time, and could result in errors if there is conflicting data in the database. If I inject a `ContactApi` instance as a constructor argument I can use a mock in tests that will pass me what I expect in the tests without touching the database. 

Furthermore, if I decide in `DupeFinder::findDupes()` to add a flag to do something related to another entity, for example returning all contact emails with the`$withEmails` flag then I'd need to add the `EmailApi` via the constructor. Some people might recoil and think this is tiresome, but by forcing the developer to realize how many related classes their class depends on I think we end up encouraging smaller classes, focused on a single responsibility.

Faster tests will make people more likely to write lots of them, covering more edge cases. Database set-up and tear down time for a single test is not huge right now, but as we have more and it becomes a major task to run all unit tests, as it is now in CiviCRM core.

It is possible to write wrapper classes that hides the static calls and inject that as a constructor argument to my class, but eventually we would have a lot of wrapper classes just to avoid the static calls.

My suggestion is to make the API use the regular OO structure of inheritance. Instead of having a base Get class, have an abstract BaseAPI class with the basic methods. Extend from this class to add new entity APIs. This API class could then be injected in the class, which would both help with testing and force you to recognize your class dependencies. If you had a class where you were injecting 20 API classes you might consider splitting it into smaller classes (which I think is good practice), but with 20 static API calls you probably won't notice this.

Here is some psuedocode to illustrate my suggested approach:

```
abstract class BaseAPI {
  /**
   * ParameterBag is just a suggestion, since it
   * would be nice to allow an OO approach to setting parameter values. It
   * would probably be similar to what we currently have for Action and
   * allow array access. It might be an option to allow an array as well.
   *
   * @param ParameterBag $params
   *
   * @return Result
   */
  public function get(ParameterBag $params) {
    // from here run the request through kernel and return result
  }

  // all the other base actions
}
```

```
class ContactAPI extends BaseAPI {
  /**
   * @param ParameterBag $params
   *
   * @return Result
   */
  public function get(ParameterBag $params) {
    // just an example if we want to enforce some behaviour unique to this entity
    $params->setOption('limit', 20);

    return parent::get($params);
  }
}
```

```
/**
 * Sample class showing usage of suggested style
 */
class DupeFinder {
  /**
   * @var ContactAPI
   */
  protected $contactApi;

  /**
   * @param ContactAPI $contactApi
   */
  public function __construct(ContactAPI $contactApi) {
    $this->contactApi = $contactApi;
  }

  /**
   * @return array
   */
  public function findDupes() {
    $dupes = array();
    $allContacts = $this->contactApi->get();

    foreach ($allContacts as $contact) {
      $params = new ParameterBag();
      $params->setValue('display_name', $contact['display_name']);
      $dupeResults = $this->contactApi->get($params);

      if ($dupeResults->getCount() > 0) {
        $dupes += $dupeResults;
      }
    }

    return $dupes;
  }
}
```

Testing the `DupeFinder` would be a lot easier then, you could just inject a mock and test a lot of different scenarios without having to setup the database.

**Comment by totten** [comment]
I'm not going to give a straight +1/-1 -- but here are some gotchas to look out for the `class ContactAPI extends BaseAPI` approach:

 * __Avoid limiting the list of actions__: APIv3 allows extensions to mix-in additional actions for existing entities. Based on grepping universe, ~15% of APIv3-related files are supplemental actions for existing core entities (like `Contact`, `Group`, or `Job`). The existence of one class/developer/extension implementing `Case.get` shouldn't preclude another class/developer/extension from implementing `Case.getdetails`.
 * __Avoid scanning the full API source-code on every page-request__: Suppose there are 200 API entities with an average of 5 API actions -- i.e. 1000 API actions total. But in a typical page-request, you might only use 10 entities or 30 actions.
   * In a typical page-request, the amount of scanning work should either be:
     * (b) Linear-time with *the number of APIs used*. (Ex: In the existing APIv4 static calling convention, a consumer calls `Participant::get()` which requires scanning `Participant.php` and maybe 1-2 parents. But it doesn't need to scan the other 199 APIs.)
     * (a) Constant-time (Ex: In Symfony's Container, a consumer calls `Container::get('service-id')`, and the container reads a cached list of service definitions. This cache is constant-time because it's only loaded once during a typical page-request.)
   * A naive implementation of `ProviderInterface` which supports the `FooAPI extends BaseAPI` convention might walk through all the classes and use reflection on all of them. If you do that, cache it.

**Comment by mickadoo** [comment]
@totten thanks for the reply

#### Avoid limiting the list of actions:

Good point, I thought about allowing developers to extend from a base API would be enough, but this doesn't allow different extensions to extend to do different things, like override an action and add a new one. This should be possible, but another things that concerns me is how to handle clashes. Eventually if you install enough extensions two of them will try to override the same thing and we'll have a class name clash. I wonder if we should rethink the design to avoid this.

I'm happy to go with whatever structure suits our needs best, my main goal is to allow developer to use the API in a non-static way because of the reasons I mentioned above.

#### Avoid scanning the full API source-code on every page-request:

The way overrides method work now depend on scanning to see if a class is declared. I agree we don't want to scan through the whole codebase to find which class to call. I'm open to options here, I feel like a broken record whenever I say "container" now, but I don't want to just say "I'm not happy with the way it is" without offering an alternative. One way to overcome the problem with clashes and scanning the directories could be:

  - Use tagged services with an attribute that specifiies which entity and action the class is responsible for. This would allow extension developers to add new actions. We would not allow overriding of existing actions.
  - For modifying existing actions behaviour use the event system, which would allow multiple extensions to modify the behaviour of each request.

This is just one possible approach, again my main aim is to replace the static calls. The reason I created this issue is to try and gauge opinion on it. If the community feels that static calls are the way to go then there's no point in me continuing with this, but I strongly feel that forcing developers to use static calls for APIv4 will be a disadvantage in the long term.

**Comment by mickadoo** [comment]
I've spent some time looking at how this might work. My main goals, influenced by the comments above were:

- Allow use of APIv4 in a non-static way
- Allow extensible actions for each entity
- Use constructor arguments to define dependencies and allow mocking in tests

## Sample Usage:

Activity.get:

```
    $activityApi = \Civi::service('activity.api');
    $params = new GetParameterBag();
    $params->addWhere('activity_type.name', '=', 'housing_support');
    $results = $activityApi->request('get', $params, FALSE); // FALSE is for $checkPermissions
```

OptionValue.get (where the API was injected as a constructor argument):

```
      $params = new GetParameterBag();
      $params->addWhere('name', '=' ,$activityType);
      $params->addWhere('option_group.name', '=', 'activity_type');
      $result = $this->optionValueApi->request('get', $params, FALSE);
```

CustomField.create

```
    $params = new ParameterBag();
    $params->set('label', 'FavColor');
    $params->set('custom_group_id', $customGroup['id']);
    $params->set('html_type', 'Text');
    $params->set('data_type', 'String');
    $customFieldApi->request('create', $params, FALSE);
```

CustomGroup.create (using array instead of ParameterBag):

```
    $customGroup = $customGroupApi->request('create', array(
      'name' => 'MyContactFields',
      'extends' => 'Contact'
    ), FALSE);
```

## Core Classes

These classes are core to the operation of the API. Their public methods are listed.

- API:
  Base class for all APIs. All entity APIs use same class, the only difference is list of available actions and entity name.
    - __construct(ApiKernel $kernel, string $entity, RequestHandlerInterface[] = array())
    - request($action, $params = NULL, $checkPermission = TRUE) : Response
    - addHandler(RequestHandlerInterface $handler)
    - getActions() : RequestHandlerInterface[]
    - getEntity() : string

- ApiRequest:
  Contains all the details for a single request, including which handler will be used for it.
    - __construct(string: Entity, RequestHandlerInterface, ParameterBag = NULL)
    - set(string, mixed)
    - get(string): mixked
    - getAll() : array
    - setCheckPermission(bool)

- Response:
  Contains the result of an API call. extends \ArrayObject.
    - __construct(array)
    - setStatusCode(int)
    - addHeader(string)
    - getStatusCode() : int
    - getHeaders() : array

- ApiKernel:
  Service to run the request lifecycle, dispatch request events
    - __construct(EventDispatcherInterface)
    - run(ApiRequest) : Response

- RequestHandlerInterface:
  Contains only the method for handling a certain request
    - __construct()
    - handle(ApiRequest $request) : Response
    - getAction() : string

## Overview
- It uses the container and tagged services extensively. APIs are added by tagging a service with the "api" tag.
- Extensions can inject additional actions and add their own APIs.
- API services tagged with "standard=true" will have the default actions injected into them.
- Since YAML is not an option, service definitions as stored in XML files.
- Conceptually the API classes are similar to controllers, with the request handlers as controller actions.
- Default request handlers can be overridden, usually when dealing with odd BAOs. For simple modification of parameters use event subscribers.

## Reasons for Change

- By using the existing APIv3 events I found that I would need to modify core code. These event listeners were becoming more complicated because they needed to be able to handle _all_ API versions. Using the same events from different version let to `if` checks where the type or class of event properties was unclear. Because of this I redid events specific to APIv4 which made it clearer what to expect and avoids modification of core code.
- The civicrm_api4() function will still work. I added the $checkPermission as an argument here as I don't feel it belongs in the $params since it alters behaviour of the function.
- Originally "Actions" were responsible for both handling requests, executuing them and storing parameters, I split this into ApiRequest to store data, and HandlerInterface for handling requests.
- Original "Actions" accepted a "Result" instance in their `_run` method and then modified it. Semantically I don't think this made sense.
- In many cases I move away from using code from CiviCRM core. By working in an extension we can remove backwards compatibilty with existing APIv3 wrappers, event listenerers etc. instead of trying to support all versions in these classes.
- Classes now define their dependencies in the contructor arguments, which makes testing easier. For example see ContactApiTest which mocks the ApiKernel.
- I believe we should support flexibility of class and namespace names and location (in directory structure). By using tagged services the name / namespace is not important.
- While it would like to have concrete methods like 'get' and 'create' in the API class the action called is a variable. There is no way to know if an action exists before calling the API.
- I could not find a way to have a single class that contains all available methods for a particular entity. That would require multiple inheritance, i.e. MyContactAPI extends BaseContactAPI, AnotherExtensionContactAPI.
- I extended the ParamterBag with the GetParameterBag because our arguments to GET requests have a complex structure.

## Unsure Points
- Whe working on this I was divided. On one hand I wanted to make modular services, each of which acted on a specic entity, e.g. ContactApi, ParticipantApi. This would limit the scope of each service and allow for easier tracking of dependencies. On the other hand a conventional HTTP API has a single entrypoint which allows access to all available endpoints, e.g. Contact.get, Participant.create. It would be possible to make the API this "god" class that allowed access to all services. Again I feel this is the conflict between an external-facing HTTP API and an internal tool for manipulating entities. The civicrm_api4() function is this "god" function that allows access to _all_ actions for _all_ entities. I think as an external facing API this is expected. I intentionally _narrowed_ the scope of the internal API class to only act on a single entity. I did this because I believe that tracking dependencies is an important way to gauge the complexity of your classes. By allowing access to every entity a class can easily become very complex without it being so obvious. On the other hand, if we require explictly injecting each specific API class for each entity then it is clearer that your class is becoming too complex. 
- I am not happy with having the RequestHandler being a property of APIRequest. I originally named the request class simply Request, and it had only scalar properties, like $action (string) and $entity (string). If using a single entrypoint this makes sense. The kernel could find the controller based on these properties and call the action. However in this API / Entity Service mashup the API class needs to keep track of the handler if we are to avoid instantiating all available handlers at runtime (which would happen if we stored the handlers in the Kernel).

## Removed

- I removed the reflection classes which fetch class comments. The only test that broke were one where a description of the $where property was returned. I would need to know more about the target use of this before putting it back.
- I removed things like entity and version from the response. I saw duplication of these parameters in a few places. I don't think the result needs to know about them.

## Suggestions

- It keeps coming up that we are trying to put two separate things into one here. Using the API to communicate between extensions seems like a good idea. However a traditional API gives access to all possible endpoints in the applicaiton. This will lead to poorly defined classes and make testing harder. I suggest separating our goals here. Allow use of APIv4 as an API to be called from external API consumers. For internal calls improve/replace the BAOs with uniform service classes. Encourage the use of the API for HTTP consumers, and encourage the use of the service classes internally.

**Comment by totten** [comment]
OK, so this PR is kinda big. ;) I haven't fully digested it yet, but I'll post a few questions/observations/suggestions from skimming:

1. Re: `GetParameterBag`
  - So it's still possible to either (a) `set` the list of `where`, `orderBy`, `select` params or (b) add items one-by-one . That's cool.
  - The parent class (`Symfony\Component\DependencyInjection\ParameterBag\ParameterBag`) has functions like `resolve()`, `resolveValue()`, and `resolveString()`. Is this stuff good or bad in our use-case? If we're mapping web-service inputs to `GetParameterBag`, might we accidentally let untrusted parties "resolve" things they shouldn't?
  - Some implementations of the "Bag" pattern mix in behavior for validations and defaults (e.g. `Symfony\Component\HttpFoundation\ParameterBag` and `CRM_Utils_OptionBag`). That seems useful.

2. Re: `\Civi::service('activity.api')`

The nice thing about `Civi::service()` is it's generic. The problem is it's generic -- there's no type-hinting. So there are a few stubs (like `Civi::dispatcher()` or `Civi::log()`) which add type-hints.

If it's actually the case that several bits of code will be looking up these services,  then a stub like `Civi\Api4::entity('activity')` might help.

3. Re: New kernel/events

> By using the existing APIv3 events I found that I would need to modify core code.  These event listeners were becoming more complicated because they needed to be able to handle all API versions.

This reflects major change in the conception of APIv4 over the years -- e.g.  earlier expectations were to focus more on more incremental change to the model/schema, and later expectations were for broader changes in the PHP syntax.  As an incremental change, it would be desirable to keep the codebases of APIv3/v4 closely aligned (doing that gives us more leverage and fewer total abstractions).  But...  given that APIv4 is doing this broad change...  I do see the appeal in forking the kernel: it's easier to iterate, and you cleanly ignore some of the more inane baggage of APIv3.

However... if we're going down this path with a total break from APIv3, then my good friend Pandora pulls a question from her box: would switching over to `HttpKernel` leave us in a better position (more recognizable to contributors) than creating a new kernel?  Then we wouldn't say an API is "like" a controller...  it would *be* a controller.

(It's perfectly fine to say "no, that doesn't make sense!" -- but I feel obliged to ask.)

4. Enumeration

For something like the API Explorer, how does one enumerate the list of entities/actions/fields?

**Comment by mickadoo** [comment]
@totten yep, the PR sort of grew. Originally I only planned to offer a non-static approach, but it grew to include the Kernel rewrite. I'm fine with splitting into smaller PRs, I don't expect this one to be merged without some discussion and iterations, but I just opened it to make it easier for anyone reading the comments to compare them.

## Re: GetParameterBag

Good catch. I unquestioningly used the only available ParameterBag that showed up in autocompletion (based on the classes available in CiviCRM core) but it is the wrong one. The `HttpFoundation\ParameterBag` more suitable to what we want.

## Re: \Civi::service('activity.api') and type-hinting

When working on this I wanted to provide a good developer experience. One area this isn't optimial is with the variable action name; following the code without Xdebug from API -> Kernel -> RequestHandlerInterface could be difficult. However for the actual services from the container I found that IDE auto-completion works well with the Symfony plugin installed. With `Civi::service()` this returned `mixed`, but you when using `Civi::container()->get('myservice')` it correctly identified the service class, and offered auto-completion on the service ID. Auto-completion was also available in the XML service files.

![autocompletion](https://user-images.githubusercontent.com/6374064/28490243-0d289a56-6ecd-11e7-857e-acec98dfd72c.gif)

We could have a stub like `Civi::api('Contact')` that would be type-hinted to return an `ApiInterface`.

## Re: New kernel/events

> Then we wouldn't say an API is "like" a controller... it would be a controller.

I would be delighted to follow convention here, it's happened a few times where I feel like I'm apeing Symfony's implementation, but tweaked to our needs. This leads back to what I said about the duality of APIv4 - is it a typical HTTP API (a single entry point to all available API endpoints in the application), or is it a modular, service-based approach to internally interacting with entities. I don't feel it can elegantly be both.

If it is a true HTTP API then the controller approach would make more sense. We would need to add some dependencies in core (as mentioned `symfony/http-kernel` isn't available to me). In this case instead of having `contact.api` we would just have a single service, `api` that handles all requests. 

My biggest worry here is that we're replacing the `civicrm_api3()` function with `$this->api->request()`, both of which have the scope to interact with _any_ entity in the application. One of them can be mocked at least, but my concern is that internally the scope of this API service is too broad to use as a dependency for classes in CiviCRM core and in extensions.

I worked with a company that used Symfony before. They used the container _everywhere_, inside services, inside entity classes.. it was easy to start with because it gives a lot of freedom, but as the application became more complicated it was less maintainable. Testing was one reason; mocking the container is possible, but makes writing tests more difficult. Another reason is that any class that has access to the container basically has the power to do anything. I found cases where validator classes were updating the values of an entity, which was after a lot of hair pulling. This is why I tried to restrict the scope of the API classes by _only_ allowing them to act on a single entity (by making `$entity` an unchangable contructor arugment of the class). It feels like a hack, but I hope my reasons are clear.

My question from this is: can we replace the API with an alternative for internal calls? I know there's been a lot of work advocating the use of the API for communicating between extensions. In one way this makes sense because before this direct BAO calls were being used, and the BAOs are not standardized. Another reason seems to be supporting backwards compatibility by modifying parameters in the API calls before passing them to the BAOs. I don't have a solution here, but I think we can definitely find one.

## Enumeration of entities, actions and fields

Entities can be fetched using the Entity.get request (all services tagged with `api` are registered in the `EntityRegister` service).

Actions are fetched with the `getActions` request. Internally this just fetches the API for that entity and gets all the actions that it has request handlers for.

Fields are fetched using the `getFields` request. Internally this uses the `SpecGatherer` class, that is responsible for getting all the request specifications (similar to the function + hook in APIv3).

**Comment by mickadoo** [comment]
I played around with the idea of using real controllers.

How the code looks:

Code calling the API:
```
    $kernel = \Civi::container()->get('civi.api.kernel');
    $request = Request::create('/contact'); // defaults to GET
    $response = $kernel->handle($request);
```

Controller:
```
class ContactController extends Controller {
  /**
   * @param Request $request
   *
   * @return Response
   */
  public function getAllAction(Request $request) {
    $contacts = $this->multiGet($request, 'Contact');

    return new Response(json_encode($contacts));
  }
}
```

Base Controller:
```
  /**
   * @param Request $request
   * @param $entity
   *
   * @return array|int
   */
  protected function multiGet(Request $request, $entity) {
    $query = new Api4SelectQuery($entity, TRUE); // todo permission

    $query->select = $request->get('select', array());
    $query->where = $request->get('where', array());
    $query->orderBy = $request->get('orderBy', array());
    $query->limit = $request->get('limit', 0);
    $query->offset = $request->get('offset', 0);

    return $query->run();
  }
```

Routes.xml:
```
    <route id="get_all_contacts" path="/contact">
        <default key="_controller">\Civi\Api4\Controller\ContactController::getAllAction</default>
    </route>
```

WIP Branch is available [here](https://github.com/mickadoo/api4/tree/http-api).

This approach is centred on providing a HTTP API. It is not designed for internal calls.

#### The good:

- Follows convention for HTTP request / response lifecycle by accepting a request generated from globals and returning a HTTP response.
- Allows usage of methods, like `PUT /contacts/1`
- Allows usage of status codes in the response
- Should be familiar to anybody who has worked on, or used, RESTful APIs
- Uses proven Symfony components
- Using real controller classes is a familiar design pattern
- By using conventional URLs it should be possible to use resources in Angular and make the API design more logical without needing to check the explorer
- The existing `RequestHandlers` (previously `Actions`) can be used without much change into the base controller class

#### The bad:

- Does not match our current view of "API requests" (entity + actions), instead we just have a collection of routes
- Is not suitable for internal use as it both requires the creation of a request object, and parsing of a JSON response object
- Disruptive, it is a different approach to how we do things and could take time to adjust
- Does not solve the problem of internal (PHP) manipulation of entities
- Adds more dependencies, more code to keep up to date
- Requires knowledge of the request lifecycle, controllers and routing

I think to resolve this we need to clearly decide what we want from APIv4, if it's an API that can be used by HTTP consumers (the Angular app being the best customer) then I think this is a good approach. The drawback is we need to work on a separate method for internal calls.

If we want to be both an external facing HTTP API and an internal tool then I think the [previous approach](https://github.com/civicrm/api4/issues/63#issuecomment-317022397) was better than this, because it's less disruptive. 

In my opinion it's better to do them separately, with the APIv4 extension purely to provide a HTTP API for external consumers. Regarding the internal calls, I think improving and standardizing the BAOs would go a long way towards this. If this isn't an option we could make a separate extension that implements this "service layer" for internal use, although I think this probably belongs in core.

**Comment by totten** [comment]
@mickadoo wow, thanks for working out the contours of an HttpKernel approach! I spent some time kicking around ideas to mitigate the bad (e.g. [this](https://gist.github.com/totten/367a63d75ea9d6ba39955bd6ea7a5c4f) could mitigate the JSON encoding/decoding waste), but I couldn't completely get around the compromises imposed by trying blend an HTTP `Response` with an in-memory/object/array `Response`. Moreover, I don't really want to turn this into a completely different project. Better to keep the focus where it's been.

I was able to get Coleman on the phone, and he pointed out that goal in his earlier draft was to support type-hinting. The `request($action, $params)` interface basically gives that up. (And I think point there: when used as an interface, the fluent command-objects provide clearer documentation/type-hinting... but they're harder to mock because the mocking framework needs to informed about multiple method calls.)

Anyway, it prompted some more reflection/comparison and reading the code a bit more. (And do excuse me -- it's a big set of code, so it takes me a few passes to get my head into it.)

Personal opinions:

 * What works in both Coleman's earlier (pre-#73) and Michael's first revision (#73):
   * Third-parties can define new entities.
   * Third-parties can define new actions for existing entities.
 * Re: Coleman's earlier approach (pre-#73)
   * What's nice (for me):
     * The command objects support auto-completion, type-hinting, fluency
     * Convention-over-configuration -- you don't have to do any wiring to define a new command object. Just make a file with a proper name.
   * What not nice (for me):
     * Can't inject any services into the CommandObjects -- construction is hardcoded as `new $class()`.
 * Re: Michael's revised approach (#73)
   * What's nice (for me):
     * Dependency injection -- the services accept injections
     * Doesn't require as many injections as a pure CommandObject approach
     * Mocking the actions is easier
   * What not nice (for me):
     * The `request($entity,$params)` contract is big. Hard to intuit the list of $actions and their $params.
     * The registration/tagging/request-handler feels a bit complicated/spread-out for me. Maybe it's 
     * Isn't there a race in calling findTaggedServiceIds() directly within `hook_civicrm_container`?

I've been kicking around [a long-ish gist](https://gist.github.com/totten/a65c4b4c74c8196adfc39810bd85e71f) which (a) allows thinner configuration and (b) allows the consumer to choose between fluency-vs-mocking.

(After getting through that... it's dawning on me that pre-#73 is actually a bit different from command-object pattern, because it couples together the parameter-object with the execution-handler. It's worth considering the trade-offs between coupling vs decoupling those. But I need a break from writing this...)

**Comment by mickadoo** [comment]
Initial comments before I go through the gist:

> Pre #73 The command objects support auto-completion, type-hinting, fluency

Pre #73 this was true for static methods in core that are [added as annotations](https://github.com/mickadoo/api4/blob/dedfe8384a6063d79c6c9f271736f828c1092e9f/Civi/Api4/AbstractEntity.php#L40). However, if you want to add a custom contact action in your extension I don't think it's possible to support code completion without modifying the `Contact` or `AbstractEntity` class in CiviCRM core, which is not an option for extension developers.

Adding code completion support for the base actions is not a problem post #73, but by opening the API class to extension by adding actions we cannot offer autocompletion for all available actions. If you think it would be useful it would not be a problem to add `get($params)`, `create($params)` etc to a `StandardAPI` class and have the core API services use this class. The reason I didn't do this is for consistency. If you want to call an action you added yourself in an extension the call would be `$contactApi->request('myaction', $params)` which is different from `$contactApi->create($params)`. 

> The goal in his earlier draft was to support type-hinting. The request($action, $params) interface basically gives that up

As mentioned above it does give this up, but only because actions are variable and can be extended. It's would be easy to add the basic action methods to a class. API classes from the container still offer some type hinting, as shown [here](https://github.com/civicrm/api4/issues/63#issuecomment-317173130).

> The request($entity,$params) contract is big

`request($action, $params)` does have a large scope, which is why I tied them to operations on a specific entity in the Api class. If we want specific service classes I agree the scope should be small, but if we want offer a drop-in repalcement for APIv3 then the scope has to stay broad. I used specific API classes tied to an entity as a compromise between application-wide scope (a-la `civicrm_api3`) and the service level scope I think we should strive towards.

> The registration/tagging/request-handler feels a bit complicated/spread-out for me.

Fair point, especially for developer without experience of the service container. The basic registration isn't too bad:

```
<service id="my_custom.api" class="Civi\Api4\Api" parent="abstract.api">
    <tag name="api" standard="true"/>
    <argument index="1">MyEntity</argument>
</service>
```

Although I can see how someone might complain that adding custom handlers is difficult. At least with the Symfony plugin you get autocompletion for this.

> Isn't there a race in calling findTaggedServiceIds() directly within hook_civicrm_container?

Yes, which is why I've proposed [this](https://github.com/civicrm/civicrm-core/pull/10756) to separate registration and handling tagged services.
